### PR TITLE
[e2e] retry stack up by default

### PR DESCRIPTION
**Pull request by pducolin**
### What does this PR do?

Change default retry strategy on infra pulumi managed failures from `noRetry` to `reUp`

### Motivation

pulumi keeps track of errors on deployment, aka [up](https://www.pulumi.com/docs/cli/commands/pulumi_up/). There are plenty of reasons why an infra setup can fail, most errors are transient and sporadic, as [this one](https://app.datadoghq.com/logs?query=%22FileNotFoundError%3A%20%5BErrno%202%5D%20No%20such%20file%20or%20directory%3A%20%27%2Frun%2Fcloud-init%2Fstatus.json%27%22%20service%3A%28gitlab-ci%20OR%20gitlab-runner%29%20&cols=host%2Cservice&fromUser=true&index=%2A&messageDisplay=inline&refresh_mode=sliding&storage=hot&stream_sort=desc&viz=timeseries&from_ts=1710692661660&to_ts=1713284661660&live=true) happening twice in one month 
image

Instead of adding known errors one by one, we can monitor them with [this dashboard](https://dddev.datadoghq.com/dashboard/r74-dyj-bm7/e2e-tests---infra-errors?fromUser=true&refresh_mode=sliding&view=spans&from_ts=1710610806537&to_ts=1713285606537&live=true).

Adding the stack name to the event metadata we send on stack up error to group by stack, and see how many retries we do per stack.

### Additional Notes

A side effect of this PR is that we would retry on any error, including rate limit ones. This will slightly increase the failure time. Checking if we can filter those out by bubbling up the error details, as currently the error is 

```
E2E INTERNAL ERROR: failed to run update: exit status 255
```

Example of [failure on Gitlab](https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/488574399#L684)

### CODE REVIEW

File: test/new-e2e/pkg/utils/infra/stack_manager.go [hunk 1]
```diff
@@ -318,26 +318,29 @@ func (sm *StackManager) getStack(ctx context.Context, name string, config runner
...
 		}
-		if retryStrategy := sm.getRetryStrategyFrom(err); retryStrategy != noRetry {
-			fmt.Fprintf(logger, "Got error that should be retried during stack up, retrying with %s strategy", retryStrategy)
-			err := sendEventToDatadog(fmt.Sprintf("[E2E] Stack %s : retrying Pulumi stack up", name), err.Error(), []string{"operation:up", fmt.Sprintf("retry:%s", retryStrategy)}, logger)
+
+		retryStrategy := sm.getRetryStrategyFrom(err)
+		err := sendEventToDatadog(fmt.Sprintf("[E2E] Stack %s : error on Pulumi stack up", name), err.Error(), []string{"operation:up", fmt.Sprintf("retry:%s", retryStrategy), fmt.Sprintf("stack:%s", stack.Name())}, logger)
+		switch retryStrategy {
+		case reUp:
+			fmt.Fprint(logger, "Got error during stack up, retring")
```
Comment (FlorentClarret) [inline test/new-e2e/pkg/utils/infra/stack_manager.go:326]: nit 

```suggestion
			fmt.Fprint(logger, "Got error during stack up, retrying")
```

File: test/new-e2e/pkg/utils/infra/stack_manager.go [hunk 2]
```diff
@@ -318,26 +318,29 @@ func (sm *StackManager) getStack(ctx context.Context, name string, config runner
 		if err == nil {
 			break
 		}
-		if retryStrategy := sm.getRetryStrategyFrom(err); retryStrategy != noRetry {
-			fmt.Fprintf(logger, "Got error that should be retried during stack up, retrying with %s strategy", retryStrategy)
-			err := sendEventToDatadog(fmt.Sprintf("[E2E] Stack %s : retrying Pulumi stack up", name), err.Error(), []string{"operation:up", fmt.Sprintf("retry:%s", retryStrategy)}, logger)
+
+		retryStrategy := sm.getRetryStrategyFrom(err)
```
Comment (FlorentClarret) [inline test/new-e2e/pkg/utils/infra/stack_manager.go:322]: Sorry if this is a stupid question but I'm confused. Shouldn't we switch the default return value to `reUp` there? https://github.com/DataDog/datadog-agent/blob/ea3b9a4214b178a94a4962a231a1c408c296f5bc/test/new-e2e/pkg/utils/infra/stack_manager.go#L398

File: test/new-e2e/pkg/utils/infra/stack_manager.go [hunk 3]
```diff
@@ -318,26 +318,29 @@ func (sm *StackManager) getStack(ctx context.Context, name string, config runner
...
+		err := sendEventToDatadog(fmt.Sprintf("[E2E] Stack %s : error on Pulumi stack up", name), err.Error(), []string{"operation:up", fmt.Sprintf("retry:%s", retryStrategy), fmt.Sprintf("stack:%s", stack.Name())}, logger)
+		switch retryStrategy {
+		case reUp:
+			fmt.Fprint(logger, "Got error during stack up, retring")
 			if err != nil {
 				fmt.Fprintf(logger, "Got error when sending event to Datadog: %v", err)
 			}
```
Comment (FlorentClarret) [inline test/new-e2e/pkg/utils/infra/stack_manager.go:329]: I think this error handling should be outside the switch

Comment (pducolin) [reply test/new-e2e/pkg/utils/infra/stack_manager.go:329]: Thank you!

File: test/new-e2e/pkg/utils/infra/stack_manager.go (hunk 2 above)
Comment (pducolin) [reply test/new-e2e/pkg/utils/infra/stack_manager.go:322]: Oh my, forgot to commit the change.Thank you for the catch!

**Review by FlorentClarret — APPROVED** [review]
Looks good to me! 

We could also clean up [this](https://github.com/DataDog/datadog-agent/blob/8adf921579c9e1df6b0e7c080fab4c1e6b6a2a87/test/new-e2e/pkg/utils/infra/retriable_errors.go#L29-L50) to remove all the `reUp` types, since they are no longer needed, but that can be done in a future PR 🙂

**Comment by pducolin** [comment]
> We could also clean up [this](https://github.com/DataDog/datadog-agent/blob/8adf921579c9e1df6b0e7c080fab4c1e6b6a2a87/test/new-e2e/pkg/utils/infra/retriable_errors.go#L29-L50) to remove all the reUp types, since they are no longer needed, but that can be done in a future PR 🙂

Oh my, it was in my original changes, stashed and forgot to pop them. Thank you, on it!

**Comment by pducolin** [comment]
Now it has the intended changes, sorry again for the mess

**Review by FlorentClarret — APPROVED** [review]
Nice!

**Pull request merged**